android/modules/database/src/java/ti/modules/titanium/database/DatabaseModule.java

```diff
@@ -53,101 +53,98 @@ public DatabaseModule()
+		}
+
+		// Fetch a path to the source database file. This is the file to be copied/installed.
+		// Throw an exception if the source database was not found.
+		Log.d(TAG, "db url is = " + url, Log.DEBUG_MODE);
+		TiUrl tiUrl = TiUrl.createProxyUrl(invocation.getSourceUrl());
+		String sourcePath = TiUrl.resolve(tiUrl.baseUrl, url, null);
+		TiBaseFile srcDb = TiFileFactory.createTitaniumFile(sourcePath, false);
+		if (srcDb.isFile() == false) {
+			if (url == null) {
+				url = "(null)";
 			}
+			throw new java.io.FileNotFoundException("Failed to find source database file: '" + url + "'");
+		}
 
-			File dbPath = ctx.getDatabasePath(name);
-
-			Log.d(TAG, "db path is = " + dbPath, Log.DEBUG_MODE);
-			Log.d(TAG, "db url is = " + url, Log.DEBUG_MODE);
-
-			TiUrl tiUrl = TiUrl.createProxyUrl(invocation.getSourceUrl());
-			String path = TiUrl.resolve(tiUrl.baseUrl, url, null);
-
-			TiBaseFile srcDb = TiFileFactory.createTitaniumFile(path, false);
-
-			Log.d(TAG, "new url is = " + url, Log.DEBUG_MODE);
-
-			if (srcDb.isFile()) {
-				InputStream is = null;
-				OutputStream os = null;
-
-				byte[] buf = new byte[8096];
-				int count = 0;
-				try {
-					is = new BufferedInputStream(srcDb.getInputStream());
-					os = new BufferedOutputStream(new FileOutputStream(dbPath));
-
-					while ((count = is.read(buf)) != -1) {
-						os.write(buf, 0, count);
-					}
-				} finally {
-					try {
-						is.close();
-					} catch (Exception ig) {
-					}
-					try {
-						os.close();
-					} catch (Exception ig) {
-					}
-				}
+		// open an empty one to get the full path and then close and delete it
+		if (name.startsWith("appdata://")) {
+			String path = name.substring(10);
+			if (path != null && path.length() > 0 && path.charAt(0) == '/') {
+				path = path.substring(1);
 			}
-			return open(name);
-
-		} catch (SQLException e) {
-			String msg = "Error installing database: " + name + " msg=" + e.getMessage();
-			Log.e(TAG, msg, e);
-			throw e;
-		} catch (IOException e) {
-			String msg = "Error installing database: " + name + " msg=" + e.getMessage();
-			Log.e(TAG, msg, e);
-			throw e;
+			File f = new File(TiFileFactory.getDataDirectory(false), path);
+			name = f.getAbsolutePath();
 		}
+
+		// Set up the destination path that the source database file will be copied to.
+		File dbPath = ctx.getDatabasePath(name);
+		Log.d(TAG, "db path is = " + dbPath, Log.DEBUG_MODE);
+
+		// Create the destination directory tree if it doesn't exist.
+		dbPath.getParentFile().mkdirs();
+
+		// Copy the source database file to the destination directory. (ie: Do the install.)
+		try (InputStream inputStream = new BufferedInputStream(srcDb.getInputStream());
+			 OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(dbPath))) {
+
+			byte[] byteBuffer = new byte[8096];
+			int byteCount = 0;
+			while ((byteCount = inputStream.read(byteBuffer)) > 0) {
+				outputStream.write(byteBuffer, 0, byteCount);
+			}
+			outputStream.flush();
```

Comment:
Close both `inputStream` and `outputStream `? `flush` wouldn't be necessary with `close` either.

Reply:
From my past experience, not all stream classes flush when closed. So, I think it's better not to assume and do it explicitly (better safe than sorry).